File: ecoepub3module/src/main/java/kr/co/smartandwise/eco_epub3_module/Drm/EBookDownloadOPMS.java
package kr.co.smartandwise.eco_epub3_module.Drm;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import android.content.Context;

import com.markany.xsync.core.XSyncCipher;
import com.markany.xsync.core.XSyncContent;

import kr.co.smartandwise.eco_epub3_module.Drm.markany.FileManager;
import kr.co.smartandwise.eco_epub3_module.Util.EBookFileUtil;

public class EBookDownloadOPMS {

	public String down(Context context, String url, EBookDownloadAsyncTask task, MarkanyBook markanyBook){
		if(url.contains("#UDID")){
			url = url.replace("#UDID", Util.getDeviceId(context));
		}

		String filePath = EBookFileUtil.getEBookStoragePath(context);
		String fileName = "OPMS_" + url.substring(url.lastIndexOf("CID=") + 4, url.indexOf("&DEVCODE=")) + ".epub";

		File dir = new File(filePath);

		if(dir.exists() == false){
			dir.mkdir();
		}

		File file = new File(filePath, fileName);
//		if(file.exists()) {
//			file.delete();
//		}

		if (url != null && url.length() > 0 && !file.exists())	{
			InputStream inputStream = null;
			FileOutputStream fileOutputStream = null; 
			byte[] buf = new byte[1024];

			try {
				HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
				connection.connect();

				int lenghtOfFile = connection.getContentLength();
				if(lenghtOfFile < 1000 && lenghtOfFile != -1){
					return null;
				}
				
				if(lenghtOfFile == -1){
					lenghtOfFile = 10 * 1024 * 1024;
				}
				
				inputStream = connection.getInputStream();

				boolean fileResult = file.createNewFile();

				if(fileResult == true) {

					fileOutputStream = new FileOutputStream(file);

					int cnt = 0;
					long readData = 0;
					long total = 0;
					
					while ((cnt = inputStream.read(buf)) != -1)	{

						total += cnt;
						// publishing the progress....
						task.onProgressUpdate((int)(total*80/lenghtOfFile));
						
						fileOutputStream.write(buf, 0, cnt);
						fileOutputStream.flush();
						readData += cnt;
					}
				}
			} catch (Exception e){
				//not handle
			}finally{
				if (fileOutputStream != null){
					try {
						fileOutputStream.close();
					} catch (IOException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
				}

				if (inputStream != null){
					try {
						inputStream.close();
					} catch (IOException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
				}
			}
		}

		String returnFileName = null;

		try {
			if (markanyBook.getFileType().equalsIgnoreCase("epub")) {
				returnFileName = FileManager.unzipOpmsMarkAny(context, file, fileName, task);
			} else if (markanyBook.getFileType().equalsIgnoreCase("pdf")) {
                java.lang.System.loadLibrary("xsync-keygen");
                XSyncContent testContent = new XSyncContent(file, Util.getDeviceId(context), XSyncCipher.RUNNING_JAVA);
                InputStream is = testContent.getInputStream(context);
                File outputFile = new File(filePath, fileName.replace(".epub", ".pdf"));
                FileOutputStream fos = new FileOutputStream(outputFile);
                byte[] buf = new byte[1024];
                int readBytes;
                int totalBytes = 0;
                while( -1 != (readBytes = is.read(buf)) )
                {
                    fos.write(buf, 0, readBytes);
                    totalBytes += readBytes;
                }
                fos.close();

				returnFileName = outputFile.getName();
			}
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		return returnFileName;
	}
}

File: app/src/main/java/eco/libros/android/ebook/download/EBookDownloadOPMS.kt
package eco.libros.android.ebook.download

import android.app.Activity
import eco.libros.android.R
import eco.libros.android.common.CustomProgressFragment
import eco.libros.android.common.utill.EBookDownloadTask
import eco.libros.android.common.utill.LibrosLog
import eco.libros.android.common.utill.LibrosUtil
import eco.libros.android.common.variable.GlobalVariable
import eco.libros.android.ui.MainActivity
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.URL
import java.util.*

class EBookDownloadOPMS {
    suspend fun down(activity: Activity, _url: String, task: CustomProgressFragment, comCode: String): String? {
        var url: String = _url
        if (url.contains("#UDID")) {
            url = url.replace("#UDID", LibrosUtil.getOriginDeviceId(activity)!!)
        }

        val uid = UUID.randomUUID().toString()

        val filePath = "${LibrosUtil.getEPUBRootPath(activity)}/${activity.resources.getString(R.string.sdcard_dir_name)}"
        val fileName = "$uid.epub"

        val dir = File(filePath)

        if (dir.exists()) {
            dir.mkdir()
        }

        val file = File(filePath, fileName)
        if (file.exists()) {
            file.delete()
        }
        task.progressTask(20)
        if (url != null && url.isNotEmpty()) {
            var inputStream: InputStream? = null
            var isComplete = true
            try {
                val uo = URL(url)

                val conn = uo.openConnection().apply {
                    readTimeout = GlobalVariable.EBOOK_TIMEOUT_SOCKET
                    connectTimeout = GlobalVariable.EBOOK_TIMEOUT_CONNECTION
                }

                var lengthOfFile = conn.contentLength

                if (lengthOfFile < 1000 && lengthOfFile != -1) {
                    return null
                }

                if (lengthOfFile == -1) {
                    lengthOfFile = 10 * 1024 * 1024
                }

                inputStream = conn.getInputStream()

                var size = inputStream.available()

                if (size <= 0) {
                    size = 1024
                }

                val buf = ByteArray(size)
                val fileResult = file.createNewFile()
                task.progressTask(30)
                if (fileResult) {
                    FileOutputStream(file).use { fos ->
                        var cnt: Int? = 0
                        var total: Long = 0
                        var readData: Long = 0
                        while (inputStream.read(buf, 0, buf.size).also { cnt = it } != -1) {
                            total += cnt?.toLong()!!
                            val i = (total * 80 / lengthOfFile).toInt()
//                            if (i > 79) {
//                                task.progressTask(79)
//                            } else {
//                                task.progressTask(i)
//                            }
                            fos.write(buf, 0, cnt!!)
                            fos.flush()
                            readData += cnt!!.toLong()
                        }
                    }
                }
                task.progressTask(80)
            } catch (e: Exception) {
                isComplete = false
                LibrosLog.print(e.toString())
                val mSocket = (activity as MainActivity).mSocket
                mSocket.emit("working_error", e.toString())
            } finally {
                try {
                    inputStream?.close()
                } catch (e: IOException) {
                    if (isComplete) {
                        var returnFileName: String? = null
                        if (fileName != null) {
                            try {
                                if (comCode != null && comCode.equals("OPMS_MARKANY", true)) {
                                    returnFileName = FileManager().unzipOpmsMarkAny(activity, file, fileName)
                                } else {
                                    returnFileName = FileManager().unzip(activity, file, fileName, uid, task)
                                }
                            } catch (e: IOException) {
                                LibrosLog.print(e.toString())
                            }
                        }
                        return returnFileName
                    } else
                        LibrosLog.print(e.toString())
                }
            }
        }

        var returnFileName: String? = null

        if (fileName != null){
            try{
                if (comCode != null && comCode.equals("OPMS_MARKANY",true)){
                    returnFileName = FileManager().unzipOpmsMarkAny(activity,file,fileName)
                } else{
                    returnFileName = FileManager().unzip(activity, file, fileName, uid, task)
                }
            } catch (e: IOException){
                LibrosLog.print(e.toString())
            }
        }
        return returnFileName
    }
}